[Property Wrappers] Support local property wrappers

hborla opened the pull request:
Support property wrappers on local variables, e.g.

```swift
@propertyWrapper
struct Wrapper<T> {
  var wrappedValue: T
}

func test() {
  @Wrapper var value = 10
}
```

hborla committed:
[Parser] Don't resolve decl references in the parser if the declaration
has a custom attribute.

hborla committed:
[NameLookup] Teach unqualified lookup to resolve backing property wrapper
and projected value references

hborla committed:
[SILGen] Teach SIlGen to emit local property wrappers

hborla committed:
[SILGen] Only use assign_by_wrapper for wrapped instance properties inside an
initializer, and for wrapped local variables.

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please test source compatibility

hborla committed:
[Property Wrappers] Make sure captures are computed for synthesized property
wrapper accessors.

hborla committed:
[Property Wrappers] Allow property wrappers on local variables.

hborla committed:
[Property Wrappers] Add a VarDecl helper method for visiting synthesized
property wrapper vars.

slavapestov commented on lib/Sema/TypeCheckStorage.cpp:
```diff
@@ -1147,8 +1147,9 @@ synthesizeTrivialGetterBody(AccessorDecl *getter, TargetImpl target,
     body.push_back(returnStmt);
   }
 
+  // Don't mark local accessors as type-checked - captures still need to be computed.
```
This seems a bit fragile because you're assuming that type checking is idempotent on already-checked AST. Would it be better to insert an explicit call to computeCaptures() inside TypeCheckDeclPrimary.cpp?

slavapestov replied:
(And yeah, computeCaptures() should be a request, since there are a few places we have to trigger it manually for synthesized code...)

hborla replied:
I tried to change this, but I discovered that "re" type checking the accessor is doing more necessary things than just computing captures.

For example, if the `wrappedValue` getter of a local property wrapper is mutating, `buildStorageReference` will generate something like for the local getter:

```
(member_ref_expr implicit type='String' decl=test.(file).Lazy.wrappedValue@test.swift:10:7 [with (substitution_map generic_signature=<Value> (substitution Value -> String))]
  (declref_expr implicit type='Lazy<String>' decl=test.(file).testLocalLazy()._value@test.swift:28:13 direct_to_storage function_ref=unapplied))
```

But it really needs to look like this:

```
(load_expr implicit type='String'
  (member_ref_expr implicit type='@lvalue String' decl=test.(file).Lazy.wrappedValue@test.swift:10:7 [with (substitution_map generic_signature=<Value> (substitution Value -> String))]
    (inout_expr implicit type='inout Lazy<String>'
      (declref_expr implicit type='@lvalue Lazy<String>' decl=test.(file).testLocalLazy()._value@test.swift:28:13 direct_to_storage function_ref=unapplied))))
```

CSApply knows how to do this, but `buildStorageReference` doesn't. Is there a reason why `buildStorageReference` is attempting to build a type-checked expression rather than relying on the constraint system to fill in types?

slavapestov replied:
Addressing this in a follow-up PR is fine. I suggest investigating why the mutating getter case works for property wrappers on members of types but not with local variables. It might be an easy fix.

slavapestov commented on include/swift/AST/NameLookup.h:
```diff
@@ -225,6 +225,8 @@ enum class UnqualifiedLookupFlags {
   /// results.
   IncludeOuterResults   = 1 << 4,
+  /// Includes property wrapper name lookup results
+  IncludePropertyWrapperResults = 1 << 5,
```
It's not entirely clear to me when you want this turned off. Do you mind adding a comment or just explaining here if it's totally obvious?

hborla replied:
I had added this flag before you removed the other implementation of unqualified lookup, so the property wrapper lookup was just performing unqualified lookup again with the prefixed names (which seemed expensive so I was limiting it to resolving decl references in the type checker). I think this can just be removed now

slavapestov commented on lib/SIL/IR/SILDeclRef.cpp:
```diff
@@ -487,7 +487,8 @@ IsSerialized_t SILDeclRef::isSerialized() const {
   // marked as @frozen.
-  if (isStoredPropertyInitializer() || isPropertyWrapperBackingInitializer()) {
+  if (isStoredPropertyInitializer() || (isPropertyWrapperBackingInitializer() &&
+      !d->getDeclContext()->isLocalContext())) {
```
Since the below code traps if `cast<NominalTypeDecl>` fails, it's arguably clearer to call !isTypeContext() instead of isLocalContext(), or to replace the cast with a dyn_cast.

hborla committed:
[Property Wrappers] Generalize a few property wrapper decl context
checks to check for type context instead of local context.

This generalization will help us implement property wrappers on
global variables, which should use the same approach of not adding
synthesized accessors to the AST and instead lazily visit them in
SILGen.

hborla committed:
[Name Lookup] Remove property wrapper name lookup flags and generalize
unqualified lookup of auxiliary decl names.

slavapestov commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -605,25 +605,21 @@ bool ASTScopeDeclConsumerForUnqualifiedLookup::consume(
+        // Check if the name matches any auxiliary decls in the same scope
+        // that are not in the AST.
+        varDecl->visitAuxiliaryDecls([&](VarDecl *auxiliaryVar) {
+          if (auxiliaryVar->getDeclContext() == varDecl->getDeclContext() &&
```
When is the DeclContext check expected to fail?

hborla replied:
Currently never, this is just a sanity check. I can remove it or make it an assertion

slavapestov reviewed (approved):
This looks good to go! I know you haven't written SILGen tests yet but I don't expect to have anything to say about those.

One more thing you might want to do in a follow-up PR is visit auxiliary decls in a couple more places, in case you weren't already planning on it:

- LookupVisibleDecls, for code completion and typo correction. Eventually LookupVisibleDecls will use ASTScope, but for now it has a duplicate implementation of unqualified lookup
- SourceEntityWalker for indexing

hborla committed:
[SILGen] Add a new CaptureEmission kind specifically for emitting
captured local variables for the assign_by_wrapper setter.

Since assign_by_wrapper will always be re-written to initialization
if the captured local variable is uninitialized, it's unnecessary
to mark the capture as an escape. This lets us support out-of-line
initialization for local property wrappers.

theblixguy commented on lib/AST/Decl.cpp:
```diff
@@ -5956,6 +5956,17 @@ VarDecl *VarDecl::getPropertyWrapperProjectionVar() const {
+
+  if (auto *projectionVar = getPropertyWrapperProjectionVar())
+    visit(projectionVar);
+}
```
Should this visit the lazy backing storage as well?

hborla replied:
Yes! I was going to try to implement local lazy in a followup PR

hborla replied:
(unless somebody else gets to it first 🙂 )

theblixguy replied:
Awesome! There’s an old WIP PR for it: https://github.com/apple/swift/pull/27600 which you may find helpful.

AnthonyLatsis replied:
Ah, that one got dropped on the floor because I ended up focusing entirely on figuring out a proper way to insert synthesized decls into a `BraceStmt` post-factum. They simply get *visited* here instead, right? I am sort of ashamed of myself for missing the point, but that's awesome; thanks, Holly!

hborla replied:
Yes, that's right - with this approach, I've added calls to `visitAuxiliaryDecls` in name lookup, the decl checker, and SILGen, so the synthesized `VarDecl`s don't need to be added to the `BraceStmt`, but they're still visited in the appropriate places.

hborla commented:
@slavapestov could you also take a look at https://github.com/apple/swift/pull/34109/commits/4bb98baf1364ce63e410c381aed7f9690dc5119f ? I want to make sure it's not a terrible idea

hborla committed:
[Test] Start to add SILGen tests for local property wrappers.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla commented:
Doesn't look like there were actually any failures in the source compat suite

hborla committed:
[Property Wrappers] Ban non-member property wrappers with observers,
just like we do for regular non-member properties.

hborla committed:
[SILGen] Don't use assign_by_wrapper for local property wrappers if
there is an initial wrapped value.

hborla commented:
@swift-ci please smoke test
